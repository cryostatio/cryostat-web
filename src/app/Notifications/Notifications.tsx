/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import { BehaviorSubject, Observable } from 'rxjs';
import { concatMap, filter, first, map } from 'rxjs/operators';
import { AlertVariant } from '@patternfly/react-core';
import { nanoid } from 'nanoid';
import { NotificationCategory } from '@app/Shared/Services/NotificationChannel.service';

export interface Notification {
  hidden?: boolean;
  read?: boolean;
  key?: string;
  title: string;
  message?: string | Error;
  category?: string;
  variant: AlertVariant;
  timestamp?: number;
}

export class Notifications {
  private readonly _notifications$: BehaviorSubject<Notification[]> = new BehaviorSubject<Notification[]>([]);
  private readonly _drawerState$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);

  constructor() {
    this._drawerState$
      .pipe(
        filter((v) => v),
        concatMap(() => this._notifications$.pipe(first()))
      )
      .subscribe((prev) =>
        this._notifications$.next(
          prev.map((n) => ({
            ...n,
            hidden: true,
          }))
        )
      );
  }

  drawerState(): Observable<boolean> {
    return this._drawerState$.asObservable();
  }

  setDrawerState(state: boolean): void {
    this._drawerState$.next(state);
  }

  notify(notification: Notification): void {
    if (!notification.key) {
      notification.key = nanoid();
    }
    notification.read = false;
    if (notification.hidden === undefined) {
      notification.hidden = this._drawerState$.getValue();
    }
    notification.timestamp = +Date.now();
    if (notification.message instanceof Error) {
      notification.message = JSON.stringify(notification.message, Object.getOwnPropertyNames(notification.message));
    } else if (typeof notification.message !== 'string') {
      notification.message = JSON.stringify(notification.message);
    }
    this._notifications$.pipe(first()).subscribe((prev) => {
      prev.unshift(notification);
      this._notifications$.next(prev);
    });
  }

  success(title: string, message?: string | Error, category?: string, hidden?: boolean): void {
    this.notify({ title, message, category, variant: AlertVariant.success, hidden });
  }

  info(title: string, message?: string | Error, category?: string, hidden?: boolean): void {
    this.notify({ title, message, category, variant: AlertVariant.info, hidden });
  }

  warning(title: string, message?: string | Error, category?: string): void {
    this.notify({ title, message, category, variant: AlertVariant.warning });
  }

  danger(title: string, message?: string | Error, category?: string): void {
    this.notify({ title, message, category, variant: AlertVariant.danger });
  }

  notifications(): Observable<Notification[]> {
    return this._notifications$.asObservable();
  }

  unreadNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(map((a) => a.filter((n) => !n.read)));
  }

  actionsNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(map((a) => a.filter((n) => this.isActionNotification(n))));
  }

  cryostatStatusNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(
      map((a) =>
        a.filter(
          (n) => (this.isWsClientActivity(n) || this.isJvmDiscovery(n)) && !Notifications.isProblemNotification(n)
        )
      )
    );
  }

  problemsNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(map((a) => a.filter(Notifications.isProblemNotification)));
  }

  setHidden(key?: string, hidden: boolean = true): void {
    if (!key) {
      return;
    }
    this._notifications$.pipe(first()).subscribe((prev) => {
      for (let n of prev) {
        if (n.key === key) {
          n.hidden = hidden;
        }
      }
      this._notifications$.next(prev);
    });
  }

  setRead(key?: string, read: boolean = true): void {
    if (!key) {
      return;
    }
    this._notifications$.pipe(first()).subscribe((prev) => {
      for (let n of prev) {
        if (n.key === key) {
          n.read = read;
        }
      }
      this._notifications$.next(prev);
    });
  }

  markAllRead(): void {
    this._notifications$.pipe(first()).subscribe((prev) => {
      for (let n of prev) {
        n.read = true;
      }
      this._notifications$.next(prev);
    });
  }

  clearAll(): void {
    this._notifications$.next([]);
  }

  private isActionNotification(n: Notification): boolean {
    return !this.isWsClientActivity(n) && !this.isJvmDiscovery(n) && !Notifications.isProblemNotification(n);
  }

  private isWsClientActivity(n: Notification): boolean {
    return n.category === NotificationCategory.WsClientActivity;
  }

  private isJvmDiscovery(n: Notification): boolean {
    return n.category === NotificationCategory.TargetJvmDiscovery;
  }

  static isProblemNotification(n: Notification): boolean {
    return n.variant === AlertVariant.warning || n.variant === AlertVariant.danger;
  }
}

const NotificationsInstance = new Notifications();

const NotificationsContext = React.createContext(NotificationsInstance);

export { NotificationsContext, NotificationsInstance };
