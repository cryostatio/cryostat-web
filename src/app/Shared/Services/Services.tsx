/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import { NotificationsInstance } from '@app/Notifications/Notifications';
import { TargetService, TargetInstance } from './Target.service';
import { TargetsService } from './Targets.service';
import { ApiService } from './Api.service';
import { NotificationChannel } from './NotificationChannel.service';
import { ReportService } from './Report.service';
import { SettingsService } from './Settings.service';
import { LoginService } from './Login.service';
import { JmxCredentials } from './JmxCredentials.service';

export interface Services {
  target: TargetService;
  targets: TargetsService;
  api: ApiService;
  jmxCredentials: JmxCredentials;
  notificationChannel: NotificationChannel;
  reports: ReportService;
  settings: SettingsService;
  login: LoginService;
}

const settings = new SettingsService();
const jmxCredentials = new JmxCredentials(() => api);
const login = new LoginService(TargetInstance, jmxCredentials, settings);
const api = new ApiService(TargetInstance, NotificationsInstance, login);
const notificationChannel = new NotificationChannel(NotificationsInstance, login);
const reports = new ReportService(login, NotificationsInstance);
const targets = new TargetsService(api, NotificationsInstance, login, notificationChannel);

const defaultServices: Services = {
  target: TargetInstance,
  targets,
  api,
  jmxCredentials,
  notificationChannel,
  reports,
  settings,
  login,
};

const ServiceContext: React.Context<Services> = React.createContext(defaultServices);

export { ServiceContext, defaultServices };
