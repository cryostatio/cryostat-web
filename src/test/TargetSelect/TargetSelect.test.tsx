/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import renderer, { act } from 'react-test-renderer';
import { cleanup, screen, waitFor, within } from '@testing-library/react';
import '@testing-library/jest-dom';
import { createMemoryHistory } from 'history';
import { of } from 'rxjs';
import { CUSTOM_TARGETS_REALM, TargetSelect } from '@app/TargetSelect/TargetSelect';
import { defaultServices, ServiceContext } from '@app/Shared/Services/Services';
import { Target } from '@app/Shared/Services/Target.service';
import { renderWithServiceContext } from '../Common';
import { NotificationsContext, NotificationsInstance } from '@app/Notifications/Notifications';

const mockFooConnectUrl = 'service:jmx:rmi://someFooUrl';
const mockBarConnectUrl = 'service:jmx:rmi://someBarUrl';
const mockBazConnectUrl = 'service:jmx:rmi://someBazUrl';

const cryostatAnnotation = {
  REALM: CUSTOM_TARGETS_REALM,
};
const mockFooTarget: Target = {
  jvmId: 'abcd',
  connectUrl: mockFooConnectUrl,
  alias: 'fooTarget',
  annotations: {
    cryostat: cryostatAnnotation,
    platform: {},
  },
};
const mockBarTarget: Target = { ...mockFooTarget, jvmId: 'efgh', connectUrl: mockBarConnectUrl, alias: 'barTarget' };
const mockBazTarget: Target = { connectUrl: mockBazConnectUrl, alias: 'bazTarget' };

const history = createMemoryHistory();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useRouteMatch: () => ({ url: history.location.pathname }),
  useHistory: () => history,
}));

jest.mock('@app/Shared/Services/Target.service', () => ({
  ...jest.requireActual('@app/Shared/Services/Target.service'), // Require actual implementation of utility functions for Target
}));

jest
  .spyOn(defaultServices.target, 'target')
  .mockReturnValueOnce(of()) // renders correctly
  .mockReturnValueOnce(of()) // contains the correct information
  .mockReturnValueOnce(of()) // renders empty state when expanded
  .mockReturnValueOnce(of(mockFooTarget)) // renders serialized target when expanded
  .mockReturnValueOnce(of(mockFooTarget)) // renders dropdown of multiple discovered targets
  .mockReturnValueOnce(of(mockFooTarget)) // creates a target if user completes modal
  .mockReturnValueOnce(of(mockFooTarget)) // deletes target when delete button clicked
  .mockReturnValueOnce(of(mockBazTarget)) // does nothing when trying to delete non-custom targets
  .mockReturnValue(of(mockFooTarget)); // other tests

jest
  .spyOn(defaultServices.targets, 'targets')
  .mockReturnValueOnce(of([mockFooTarget])) // renders correctly
  .mockReturnValueOnce(of([mockFooTarget])) // contains the correct information
  .mockReturnValueOnce(of([mockFooTarget])) // renders empty state when expanded
  .mockReturnValueOnce(of([mockFooTarget])) // renders serialized target when expanded
  .mockReturnValueOnce(of([mockFooTarget, mockBarTarget])) // renders dropdown of multiple discovered targets
  .mockReturnValue(of([mockFooTarget, mockBarTarget])); // other tests

jest.spyOn(defaultServices.api, 'createTarget').mockReturnValue(of(true));

jest.spyOn(defaultServices.api, 'deleteTarget').mockReturnValue(of(true));

jest
  .spyOn(defaultServices.settings, 'deletionDialogsEnabledFor')
  .mockReturnValueOnce(true) // renders correctly
  .mockReturnValueOnce(true) // renders empty state when expanded
  .mockReturnValueOnce(true) // renders serialized target when expanded
  .mockReturnValueOnce(true) // contains the correct information
  .mockReturnValueOnce(true) // renders dropdown of multiple discovered targets
  .mockReturnValueOnce(true) // creates a target if user completes modal
  .mockReturnValueOnce(false) // deletes target when delete button clicked
  .mockReturnValue(true); // other tests

describe('<TargetSelect />', () => {
  afterEach(cleanup);

  it('renders correctly', async () => {
    let tree;
    await act(async () => {
      tree = renderer.create(
        <ServiceContext.Provider value={defaultServices}>
          <NotificationsContext.Provider value={NotificationsInstance}>
            <TargetSelect />
          </NotificationsContext.Provider>
        </ServiceContext.Provider>
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('contains the correct information', async () => {
    renderWithServiceContext(<TargetSelect />);

    expect(screen.getByText('Target JVM')).toBeInTheDocument();
    expect(screen.getByText(`Select target...`)).toBeInTheDocument();

    expect(screen.getByLabelText('Create target')).toBeInTheDocument();
    expect(screen.getByLabelText('Delete target')).toBeInTheDocument();
    expect(screen.getByLabelText('Options menu')).toBeInTheDocument();
  });

  it('renders empty state when expanded', async () => {
    const { container, user } = renderWithServiceContext(<TargetSelect />);

    expect(screen.getByText('Select target...')).toBeInTheDocument();

    const expandButton = screen.getByLabelText('Details');
    await user.click(expandButton);

    const articleElement = container.querySelector('article');
    expect(articleElement).toBeInTheDocument();
    expect(articleElement).toBeVisible();
    expect(screen.getByText(`No target selected`)).toBeInTheDocument();
    expect(screen.getByText(`No target selected`)).toBeVisible();
    expect(screen.getByText(`To view this content, select a JVM target.`)).toBeInTheDocument();
    expect(screen.getByText(`To view this content, select a JVM target.`)).toBeVisible();
  });

  it('renders serialized target when expanded', async () => {
    const { container, user } = renderWithServiceContext(<TargetSelect />);

    const expandButton = screen.getByLabelText('Details');
    await user.click(expandButton);

    const codeElement = container.querySelector('code');
    expect(codeElement).toBeTruthy();
    expect(codeElement).toBeInTheDocument();
    expect(codeElement).toBeVisible();
    expect(codeElement!.textContent).toBeTruthy();
    const codeContent = codeElement!.textContent!.replace(/[\s]/g, '');
    expect(codeContent).toEqual(JSON.stringify(mockFooTarget, null, 0).replace(/[\s]/g, ''));
  });

  it('renders dropdown of multiple discovered targets', async () => {
    const { user } = renderWithServiceContext(<TargetSelect />);

    expect(screen.getByText(`fooTarget`)).toBeInTheDocument();

    await user.click(screen.getByLabelText('Options menu'));
    expect(screen.getByLabelText('Select Target')).toBeInTheDocument();
    expect(screen.getByText(`Select target...`)).toBeInTheDocument();
    expect(screen.getByText(`fooTarget (service:jmx:rmi://someFooUrl)`)).toBeInTheDocument();
    expect(screen.getByText(`barTarget (service:jmx:rmi://someBarUrl)`)).toBeInTheDocument();
    expect(screen.getByText('2')).toBeInTheDocument(); // Number of discoverable targets
  });

  it('creates a target if user completes modal', async () => {
    const createTargetRequestSpy = jest.spyOn(defaultServices.api, 'createTarget');
    const { user } = renderWithServiceContext(<TargetSelect />);

    const createButton = screen.getByLabelText('Create target');
    await user.click(createButton);

    const textBoxes = screen.getAllByRole('textbox');

    await user.type(textBoxes[0], 'service:jmx:rmi://someBazUrl');
    await user.type(textBoxes[1], 'bazTarget');

    await user.click(screen.getByText('Create'));

    expect(createTargetRequestSpy).toBeCalledTimes(1);
    expect(createTargetRequestSpy).toBeCalledWith(mockBazTarget);
  });

  it('deletes target when delete button clicked', async () => {
    const { user } = renderWithServiceContext(<TargetSelect />);

    const deleteButton = screen.getByLabelText('Delete target');
    await waitFor(() => expect(deleteButton).not.toBeDisabled());

    await user.click(deleteButton);

    const deleteTargetRequestSpy = jest.spyOn(defaultServices.api, 'deleteTarget');
    expect(deleteTargetRequestSpy).toBeCalledTimes(1);
    expect(deleteTargetRequestSpy).toBeCalledWith(mockFooTarget);
  });

  it('does nothing when trying to delete non-custom targets', async () => {
    const deleteTargetRequestSpy = jest.spyOn(defaultServices.api, 'deleteTarget');
    const { user } = renderWithServiceContext(<TargetSelect />);

    const deleteButton = screen.getByLabelText('Delete target');
    await user.click(deleteButton);

    expect(deleteTargetRequestSpy).toBeCalledTimes(0);
    expect(deleteButton).toBeDisabled();
  });

  it('deletes target when warning modal is accepted', async () => {
    const deleteTargetRequestSpy = jest.spyOn(defaultServices.api, 'deleteTarget');
    const { user } = renderWithServiceContext(<TargetSelect />);

    const deleteButton = screen.getByLabelText('Delete target');
    expect(deleteButton).toBeInTheDocument();
    expect(deleteButton).toBeVisible();

    await waitFor(() => expect(deleteButton).not.toBeDisabled());

    await user.click(deleteButton);

    const warningDialog = await screen.findByRole('dialog');
    expect(warningDialog).toBeInTheDocument();
    expect(warningDialog).toBeVisible();

    const acceptDeleteButton = within(warningDialog).getByText('Delete');
    expect(acceptDeleteButton).toBeInTheDocument();
    expect(acceptDeleteButton).toBeVisible();

    await user.click(acceptDeleteButton);

    expect(deleteTargetRequestSpy).toBeCalledTimes(1);
    expect(deleteTargetRequestSpy).toBeCalledWith(mockFooTarget);
  });

  it('does not create a target if user leaves connectUrl empty', async () => {
    const createTargetRequestSpy = jest.spyOn(defaultServices.api, 'createTarget');
    const { user } = renderWithServiceContext(<TargetSelect />);

    const createButton = screen.getByLabelText('Create target');
    await user.click(createButton);

    const confirmButton = screen.getByText('Create');
    expect(confirmButton).toBeInTheDocument();
    expect(confirmButton).toBeVisible();
    expect(confirmButton).toBeDisabled();

    const connectUrlInput = screen.getByLabelText('Connection URL') as HTMLInputElement;
    expect(connectUrlInput).toBeInTheDocument();
    expect(connectUrlInput).toBeVisible();

    connectUrlInput.focus();

    await user.keyboard('{Enter}');
    expect(createTargetRequestSpy).toBeCalledTimes(0);
  });
});
