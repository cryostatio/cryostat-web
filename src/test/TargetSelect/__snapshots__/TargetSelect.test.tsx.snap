// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TargetSelect /> renders correctly 1`] = `
<article
  className="pf-c-card pf-m-compact pf-m-rounded"
  data-ouia-component-id="OUIA-Generated-Card-1"
  data-ouia-component-type="PF4/Card"
  data-ouia-safe={true}
  id=""
>
  <div
    className="pf-c-card__header"
  >
    <div
      className="pf-c-card__header-toggle"
    >
      <button
        aria-disabled={false}
        aria-expanded={false}
        aria-label="Details"
        aria-labelledby="expandable-card-title target-select-expand-button"
        className="pf-c-button pf-m-plain"
        data-ouia-component-id="OUIA-Generated-Button-plain-1"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        disabled={false}
        id="target-select-expand-button"
        onClick={[Function]}
        role={null}
        type="button"
      >
        <span
          className="pf-c-card__header-toggle-icon"
        >
          <svg
            aria-hidden="true"
            aria-labelledby={null}
            fill="currentColor"
            height="1em"
            role="img"
            style={
              Object {
                "verticalAlign": "-0.125em",
              }
            }
            viewBox="0 0 256 512"
            width="1em"
          >
            <path
              d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      className="pf-c-card__title"
    >
      Target JVM
    </div>
    <div
      className="pf-c-card__actions"
    >
      <button
        aria-disabled={false}
        aria-label="Create target"
        className="pf-c-button pf-m-control"
        data-ouia-component-id="OUIA-Generated-Button-control-1"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        disabled={false}
        onClick={[Function]}
        role={null}
        type="button"
      >
        <span
          className="pf-c-button__icon pf-m-start"
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            fill="currentColor"
            height="1em"
            role="img"
            style={
              Object {
                "verticalAlign": "-0.125em",
              }
            }
            viewBox="0 0 512 512"
            width="1em"
          >
            <path
              d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
            />
          </svg>
        </span>
      </button>
      <button
        aria-disabled={true}
        aria-label="Delete target"
        className="pf-c-button pf-m-control pf-m-disabled"
        data-ouia-component-id="OUIA-Generated-Button-control-2"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        disabled={true}
        onClick={[Function]}
        role={null}
        tabIndex={null}
        type="button"
      >
        <span
          className="pf-c-button__icon pf-m-start"
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            fill="currentColor"
            height="1em"
            role="img"
            style={
              Object {
                "verticalAlign": "-0.125em",
              }
            }
            viewBox="0 0 448 512"
            width="1em"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    className="pf-c-card__body"
  >
    <div
      className="pf-c-select"
      data-ouia-component-id="OUIA-Generated-Select-single-1"
      data-ouia-component-type="PF4/Select"
      data-ouia-safe={true}
    >
      <button
        aria-expanded={false}
        aria-haspopup="listbox"
        aria-label="Options menu"
        aria-labelledby=" pf-select-toggle-id-2"
        className="pf-c-select__toggle"
        disabled={false}
        id="pf-select-toggle-id-2"
        onClick={[Function]}
        onKeyDown={[Function]}
        type="button"
      >
        <div
          className="pf-c-select__toggle-wrapper"
        >
          <span
            className="pf-c-select__toggle-icon"
          >
            <svg
              aria-hidden={true}
              aria-labelledby={null}
              fill="currentColor"
              height="1em"
              role="img"
              style={
                Object {
                  "verticalAlign": "-0.125em",
                }
              }
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M22.3,576.00143 L1001.8,576.00143 C1013.2,576.00143 1020.3,570.90143 1023.1,560.50143 C1025.6,550.20143 1022.9,537.30143 1014.4,522.10143 C1014.4,522.10143 874.3,232.50143 863.8,211.00143 C856,191.90143 841.4,192.000713 841.4,192.000713 L185.6,192.000713 C185.6,192.000713 169,191.60143 161.8,209.70143 C151.3,232.30143 9.7,522.00143 9.7,522.00143 C1.2,537.10143 -1.7,550.00143 1,560.40143 C3.7,570.80143 10.8,576.00143 22.3,576.00143 Z M1013.1,650.90143 C1005.8,643.60143 997.2,640.00143 987.4,640.00143 L36.6,640.00143 C26.6,640.00143 18.2,643.60143 10.9,650.90143 C3.6,658.20143 0,666.80143 0,676.60143 L0,795.50143 C0,805.50143 3.6,813.90143 10.7,821.20143 C18,828.40143 26.4,832.00143 36.4,832.00143 L987.4,832.00143 C997.4,832.00143 1005.8,828.40143 1013.1,821.10143 C1020.4,813.80143 1024,805.40143 1024,795.40143 L1024,676.60143 C1024,666.60143 1020.4,658.20143 1013.1,650.90143 Z M128,768.00143 L64,768.00143 L64,704.00143 L128,704.00143 L128,768.00143 Z M256,768.00143 L192,768.00143 L192,704.00143 L256,704.00143 L256,768.00143 Z M384,768.00143 L320,768.00143 L320,704.00143 L384,704.00143 L384,768.00143 Z M512,768.00143 L448,768.00143 L448,704.00143 L512,704.00143 L512,768.00143 Z M960,768.00143 L896,768.00143 L896,704.00143 L960,704.00143 L960,768.00143 Z"
              />
            </svg>
          </span>
          <span
            className="pf-c-select__toggle-text"
          >
            Select target...
          </span>
        </div>
        <span
          className="pf-c-select__toggle-arrow"
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            fill="currentColor"
            height="1em"
            role="img"
            style={
              Object {
                "verticalAlign": "-0.125em",
              }
            }
            viewBox="0 0 320 512"
            width="1em"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</article>
`;
